LiS/man/man8/timetst.8.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: timetst.8.man,v $
.\" Revision 0.9.2.4  2008-04-29 08:33:20  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.3  2006/04/05 10:05:02  brian
.\" - get references in right place
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:41  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH TIMETST 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ System Admin."
.SH NAME
.B timetst
\- timer tests for LiS STREAMS package
.\"
.\"
.SH SYNOPSIS
.PP
.\".HP 12
.\"\fBtimetst\fR [\fIcommand\fR]
.\".PD 0
.HP 12
\fBtimetst\fR [\fIoptions\fR]
.PD 0
.HP 12
\fBtimetst\fR {\fB-h\fR,\fB--help\fR}
.HP 12
\fBtimetst\fR {\fB-V\fR,\fB--version\fR}
.HP 12
\fBtimetst\fR {\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B timetst
is a test program for testing the LiS STREAMS package.
.\"
.\"
.\".SH COMMANDS
.\".PP
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-v\fP, \fB--verbose\fP [\fILEVEL\fP]
increase verbosity or set to \fILEVEL\fP.  The default is one (1).
This option may be repeated.
.TP
\fB-q\fP, \fB--quiet\fP
Suppress normal output (equivalent to \fB--verbose\fP=\fI0\fP).
.TP
\fB-h\fP, \fB--help\fP, \fB-?\fP, \fB--?\fP
print a usage message and exit.
.TP
\fB-V\fP, \fB--version\fP
print the version and exit.
.TP
\fB-C\fP, \fB--copying\fP
print copying permission and exit.
.\"
.\"
.\".SH DIAGNOSTICS
.\".PP
.\"
.\"
.SH BUGS
.PP
.B timetst
has no known bugs.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR oc (1),
.BR rwf (1),
.BR pafd (1),
.BR psfd (1),
.BR lis_test (8),
.BR lis_test2 (8),
.BR test-streams (8),
.BR strtst (8),
.BR thrtst (8),
.BR STREAMS (4),
.IR "Linux STREAMS Installation and Reference Manual" .
.\"
.\"
.SH COMPATIBILITY
.PP
.BR timetst
is compatible with
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH CONFORMANCE
.PP
.BR timetst
does not conform to any standard.
.B timetst
tests conformance of
.BR STREAMS (4)
to the specifications listed on the
.BR STREAMS (4)
manual page.
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH HISTORY
.PP
.B timetst
is provided with the
.I @PACKAGE_TITLE@
package.
.[
$LIST$
.]
.TI
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-05-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Deprecated sub-packages have been excluded from the master package
     release.  The excluded sub-packages are:

         - `sctp-0.2.27.20090405'
         - `iperf-2.0.8.20090405'
         - `netperf-2.3.7.20090405'
         - `osr61-0.9.2.3.20090405'

   - Documentation license upgraded to GFDL 1.3.

   - Updated tool chain to `m4-1.4.13', `automake-1.10.2',
     `bison-2.4.1', `libtool-2.2.6a', `texinfo-4.13a'.

   - Addition of the following sub-packages:

         - `stratm-0.9.2.1.20090405'

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant feature updates were made to:

         - `strchan-0.9.2.4.20090405'
         - `strsock-0.9.2.4.20090405'
         - `strnsl-0.9.2.4.20090405'
         - `sigtran-0.9.2.4.20090405'

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel2009XXXX_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the

LiS/util/timetst.c
/*****************************************************************************

 @(#) $RCSfile: timetst.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:27 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:27 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: timetst.c,v $
 Revision 1.1.1.2.4.9  2008-04-29 08:33:27  brian
 - update headers for Affero release

 Revision 1.1.1.2.4.8  2007/08/14 10:47:31  brian
 - GPLv3 header update

 Revision 1.1.1.2.4.5  2005/07/13 12:01:23  brian
 - working up compat and check pass (finally lindented LiS)

 Revision 1.1.1.2.4.3  2005/05/14 08:35:14  brian
 - copyright header correction

 *****************************************************************************/

#ident "@(#) $RCSfile: timetst.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:27 $"

static char const ident[] =
    "$RCSfile: timetst.c,v $ $Name:  $($Revision: 1.1.1.2.4.9 $) $Date: 2008-04-29 08:33:27 $";

/*
 * Copyright 1997 David Grothe, Gcom, Inc <dave@gcom.com>
 */

#define	inline			/* make disappear */

#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>
#include <string.h>		/* for strerror */
#ifdef LINUX
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/ioctl.h>
#endif
#ifdef SYS_QNX
#include <ioctl.h>
#endif

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include <sys/stropts.h>

#ifdef LINUX
#include <sys/LiS/linuxio.h>
#else
#ifdef SYS_QNX
#include <sys/LiS/qnxio.h>
#else
#include <sys/LiS/usrio.h>
#endif
#endif

#include <sys/LiS/loop.h>	/* an odd place for this file */

/************************************************************************
*                      File Names                                       *
************************************************************************/
#ifdef LINUX
#define	LOOP_1		"/dev/loop.1"
#define	LOOP_2		"/dev/loop.2"
#define LOOP_CLONE	"/dev/loop_clone"
#else
#ifdef QNX
#define	LOOP_1		"/dev/gcom/loop.1"
#define	LOOP_2		"/dev/gcom/loop.2"
#define LOOP_CLONE	"/dev/gcom/loop_clone"
#else
#define	LOOP_1		"loop.1"
#define	LOOP_2		"loop.2"
#define LOOP_CLONE	"loop_clone"
#endif
#endif

#define	ALL_DEBUG_BITS	( ~ ((unsigned long long) 0) )

#ifdef DIRECT_USER		/* tie-in to cmn_err */
#define	ENO(neg_e)	(-(neg_e))
#else
#define	ENO(dmy)	(errno)
#endif

/************************************************************************
*                           Storage                                     *
************************************************************************/

unsigned long long debug_mask;
char buf[1000];				/* general purpose */
char rdbuf[1000];			/* for reading */
long iter_cnt = 100000;			/* default iteration count */
char ctlbuf[1000];			/* control messages */
char rdctlbuf[1000];			/* control messages */

struct strbuf wr_ctl = { 0, 0, ctlbuf };
struct strbuf wr_dta = { 0, 0, buf };
struct strbuf rd_ctl = { 0, 0, rdctlbuf };
struct strbuf rd_dta = { 0, 0, rdbuf };

int output = 1;

extern void make_nodes(void);

/************************************************************************
*                        Histogram Maintenance                          *
************************************************************************/

int latency_opt = 0;			/* option to measure latency */

typedef struct {
	int micro_secs;
	unsigned counter;

} histogram_bucket_t;

#define	HISTOGRAM_INIT	{ \
			    {1, 0}, \
			    {2, 0}, \
			    {3, 0}, \
			    {4, 0}, \
			    {5, 0}, \
			    {6, 0}, \
			    {7, 0}, \
			    {8, 0}, \
			    {9, 0}, \
			    {10, 0}, \
			    {20, 0}, \
			    {30, 0}, \
			    {40, 0}, \
			    {50, 0}, \
			    {60, 0}, \
			    {70, 0}, \
			    {80, 0}, \
			    {90, 0}, \
			    {100, 0}, \
			    {200, 0}, \
			    {300, 0}, \
			    {400, 0}, \
			    {500, 0}, \
			    {600, 0}, \
			    {700, 0}, \
			    {800, 0}, \
			    {900, 0}, \
			    {1000, 0}, \
			    {2000, 0}, \
			    {3000, 0}, \
			    {4000, 0}, \
			    {5000, 0}, \
			    {6000, 0}, \
			    {7000, 0}, \
			    {8000, 0}, \
			    {9000, 0}, \
			    {10000, 0}, \
			    {20000, 0}, \
			    {30000, 0}, \
			    {40000, 0}, \
			    {50000, 0}, \
			    {60000, 0}, \
			    {70000, 0}, \
			    {80000, 0}, \
			    {90000, 0}, \
			    {100000, 0}, \
			    {200000, 0}, \
			    {300000, 0}, \
			    {400000, 0}, \
			    {500000, 0}, \
			    {600000, 0}, \
			    {700000, 0}, \
			    {800000, 0}, \
			    {900000, 0}, \
			    {1000000, 0}, \
			    {2000000, 0}, \
			    {3000000, 0}, \
			    {4000000, 0}, \
			    {5000000, 0}, \
			    {6000000, 0}, \
			    {7000000, 0}, \
			    {8000000, 0}, \
			    {9000000, 0}, \
			    {0, 0} \
			}

histogram_bucket_t hist_init[] = HISTOGRAM_INIT;
histogram_bucket_t hist[] = HISTOGRAM_INIT;

/*
 * Initialize the 'hist' buffer.
 */
void
init_hist(histogram_bucket_t * h)
{
	memcpy(h, hist_init, sizeof(hist_init));
}

/*
 * Enter a time interval into a histogram bucket
 */
void
enter_hist(histogram_bucket_t * h, int interval)
{
	for (; h->micro_secs != 0; h++) {
		if (interval <= h->micro_secs)
			break;
	}

	h->counter++;
}

/*
 * Print out the histogram
 */
void
print_hist(histogram_bucket_t * h, int minimum)
{
	histogram_bucket_t *p = h;
	int max_counter = 0;

	for (; p->micro_secs != 0; p++) {
		if (p->counter > max_counter)
			max_counter = p->counter;
	}

	for (p = h; p->micro_secs != 0; p++) {
		if (p->counter >= minimum)
			if (output)
				printf("%8d %12u\n", p->micro_secs, p->counter);
	}

	if (p->counter != 0)
		if (output)
			printf("%8s %12u\n", "Larger", p->counter);

}

/************************************************************************
*                           register_drivers                            *
*************************************************************************
*									*
* Register the drivers that we are going to use in the test with	*
* streams.								*
*									*
************************************************************************/
void
register_drivers(void)
{
#if	!defined(LINUX) && !defined(QNX)
	port_init();		/* stream head init routine */
#endif

}				/* register_drivers */

/************************************************************************
*                           timing_test                                 *
*************************************************************************
*									*
* Time reading and writing messages through streams.			*
*									*
************************************************************************/
void
timing_test(void)
{
	time_t time_on;
	time_t et;
	long i;
	int fd1;
	int fd2;
	int rslt;
	int lgth;
	int arg;
	int flags = MSG_ANY;
	int rband = 0;
	struct strioctl ioc;
	struct timeval *tp = NULL;
	struct timeval *rtp = NULL;
	int xmit_time;
	int rcv_time;
	int interval;
	struct timeval x;

	if (output)
		printf("\nBegin timing test\n");

	fd1 = user_open(LOOP_1, O_RDWR, 0);
	if (fd1 < 0) {
		if (output)
			printf("loop.1: %s\n", strerror(-fd1));
		return;
	}

	fd2 = user_open(LOOP_2, O_RDWR, 0);
	if (fd2 < 0) {
		if (output)
			printf("loop.2: %s\n", strerror(-fd2));
		return;
	}

	ioc.ic_cmd = LOOP_SET;
	ioc.ic_timout = 10;
	ioc.ic_len = sizeof(int);
	ioc.ic_dp = (char *) &arg;

	arg = 2;
	rslt = user_ioctl(fd1, I_STR, &ioc);
	if (rslt < 0) {
		if (output)
			printf("loop.1: ioctl LOOP_SET: %s\n", strerror(-rslt));
	}

	if (latency_opt) {
		tp = (struct timeval *) buf;
		strcpy(&buf[sizeof(*tp)], "Data to send down the file");
		lgth = sizeof(*tp) + strlen(&buf[sizeof(*tp)]);
		rtp = (struct timeval *) rdbuf;
	} else {
		strcpy(buf, "Data to send down the file");
		lgth = strlen(buf);
	}

	if (output) {
		printf("Time test:  write %d bytes, read/write and service queue: ", lgth);
		fflush(stdout);
	}

	sync();			/* do file sync now rather than in the middle of the test. */
	init_hist(hist);
	time_on = time(NULL);
#if 1
	for (i = 0; i < iter_cnt; i++) {
		if (latency_opt)
			gettimeofday(tp, NULL);

		rslt = user_write(fd1, buf, lgth);

		if (rslt != lgth) {
			if (output) {
				if (rslt < 0)
					printf("loop.1: write: %s\n", strerror(-rslt));
				else
					printf("loop.1:  write returned %d, expected %d\n", rslt,
					       lgth);
			}
			break;
		}

		rslt = user_read(fd2, rdbuf, lgth);
		if (rslt != lgth) {
			if (output) {
				if (rslt < 0)
					printf("loop.2: read: %s\n", strerror(-rslt));
				else
					printf("loop.2:  read returned %d, expected %d\n", rslt,
					       lgth);
			}
			break;
		}

		if (latency_opt) {
			gettimeofday(&x, NULL);	/* current time */
			xmit_time = rtp->tv_sec * 1000000 + rtp->tv_usec;
			rcv_time = x.tv_sec * 1000000 + x.tv_usec;
			interval = rcv_time - xmit_time;
			enter_hist(hist, interval);
		}
	}
#endif
	et = (time(NULL) - time_on) * 1000000;	/* time in usecs */

	if (output)
		printf("%ld micro-secs\n", et / iter_cnt);
	if (latency_opt)
		print_hist(hist, 1000);
	if (output)
		printf("\n");

	if (output)
		printf("%ld micro-secs\n", et / iter_cnt);

	ioc.ic_cmd = LOOP_PUTNXT;	/* use putnxt rather then svcq */
	ioc.ic_len = 0;
	rslt = user_ioctl(fd1, I_STR, &ioc);
	if (rslt < 0) {
		if (output)
			printf("loop.1: ioctl LOOP_PUTNXT: %s\n", strerror(-rslt));
	}

	if (output)
		printf("Time test:  write %d bytes, read/write w/o service queue: ", lgth);
	if (output)
		fflush(stdout);
	init_hist(hist);
	sync();
	time_on = time(NULL);

	for (i = 0; i < iter_cnt; i++) {
		if (latency_opt)
			gettimeofday(tp, NULL);

		rslt = user_write(fd1, buf, lgth);

		if (rslt != lgth) {
			if (output) {
				if (rslt < 0)
					printf("loop.1: write: %s\n", strerror(-rslt));
				else
					printf("loop.1:  write returned %d, expected %d\n", rslt,
					       lgth);
			}
			break;
		}

		rslt = user_read(fd2, rdbuf, lgth);
		if (rslt != lgth) {
			if (output) {
				if (rslt < 0)
					printf("loop.2: read: %s\n", strerror(-rslt));
				else
					printf("loop.2:  read returned %d, expected %d\n", rslt,
					       lgth);
			}
			break;
		}

		if (latency_opt) {
			gettimeofday(&x, NULL);	/* current time */
			xmit_time = rtp->tv_sec * 1000000 + rtp->tv_usec;
			rcv_time = x.tv_sec * 1000000 + x.tv_usec;
			interval = rcv_time - xmit_time;
			enter_hist(hist, interval);
		}
	}

	et = (time(NULL) - time_on) * 1000000;	/* time in usecs */

	if (output)
		printf("%ld micro-secs\n", et / iter_cnt);
	if (latency_opt)
		print_hist(hist, 1000);
	if (output)
		printf("\n");

	ioc.ic_cmd = LOOP_PUTNXT;	/* use putnxt rather then svcq */
	ioc.ic_len = 0;
	rslt = user_ioctl(fd1, I_STR, &ioc);
	if (rslt < 0) {
		if (output)
			printf("loop.1: ioctl LOOP_PUTNXT: %s\n", strerror(-rslt));
	}

	if (output)
		printf("Time test:  write %d bytes, getmsg/putmsg w/o service queue: ", lgth);
	fflush(stdout);
	init_hist(hist);
	sync();
	time_on = time(NULL);

	rd_ctl.maxlen = sizeof(rdctlbuf);
	rd_dta.maxlen = sizeof(rdbuf);
	wr_ctl.len = -1;
	wr_dta.len = lgth;
	for (i = 0; i < iter_cnt; i++) {
		if (latency_opt)
			gettimeofday(tp, NULL);

		rslt = user_putpmsg(fd1, &wr_ctl, &wr_dta, 0, MSG_BAND);

		if (rslt < 0) {
			if (output)
				printf("loop.1: putmsg: %s\n", strerror(-rslt));
			break;
		}

		rd_ctl.len = -1;
		rd_dta.len = -1;
		flags = MSG_ANY;

		rslt = user_getpmsg(fd2, &rd_ctl, &rd_dta, &rband, &flags);

		if (rslt < 0) {
			if (output)
				printf("loop.2: getmsg: %s\n", strerror(-rslt));
			break;
		} else if (rd_dta.len != lgth || rd_ctl.len > 0) {
			if (output) {
				printf("expected rd_ctl.len = %d, got %d\n", -1, rd_ctl.len);
				printf("expected rd_dta.len = %d, got %d\n", lgth, rd_dta.len);
			}
		}

		if (latency_opt) {
			gettimeofday(&x, NULL);	/* current time */
			xmit_time = rtp->tv_sec * 1000000 + rtp->tv_usec;
			rcv_time = x.tv_sec * 1000000 + x.tv_usec;
			interval = rcv_time - xmit_time;
			enter_hist(hist, interval);
		}
	}

	et = (time(NULL) - time_on) * 1000000;	/* time in usecs */

	if (output)
		printf("%ld micro-secs\n", et / iter_cnt);
	if (latency_opt)
		print_hist(hist, 1000);
	if (output)
		printf("\n");

	if (output)
		printf("%ld micro-secs\n", et / iter_cnt);

	user_close(fd1);
	user_close(fd2);

}				/* timing_test */

/************************************************************************
*                          set_debug_mask                               *
*************************************************************************
*									*
* Use stream ioctl to set the debug mask for streams.			*
*									*
************************************************************************/
void
set_debug_mask(unsigned long long msk)
{
	int fd;
	int rslt;
	unsigned long mask1 = (unsigned long) (msk & 0xFFFFFFFF);
	unsigned long mask2 = (msk >> 32);

	fd = user_open(LOOP_1, O_RDWR, 0);
	if (fd < 0) {
		if (output)
			printf("loop.1: %s\n", strerror(ENO(fd)));
		exit(1);
	}

	rslt = user_ioctl(fd, I_LIS_SDBGMSK, mask1);
	if (rslt < 0) {
		if (output)
			printf("loop.1: I_LIS_SDBGMSK: %s\n", strerror(ENO(rslt)));
		exit(1);
	}

	rslt = user_ioctl(fd, I_LIS_SDBGMSK2, mask2);
	if (output)
		printf("\nSTREAMS debug mask set to 0x%08lx%08lx\n", mask2, mask1);

	user_close(fd);

}				/* set_debug_mask */

/************************************************************************
*                              main                                     *
************************************************************************/

void
copying(int argc, char *argv[])
{
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
\n\
Copyright (c) 2003-2007  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 1997       David Grothe, Gcom, Inc <dave@gcom.com>\n\
\n\
All Rights Reserved.\n\
\n\
This program is free software;  you can  redistribute  it and/or modify it under\n\
the terms of the GNU General  Public License as  published by the  Free Software\n\
Foundation; version 3 of the License.\n\
\n\
This program is distributed in the hope that it will be  useful, but WITHOUT ANY\n\
WARRANTY;  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
\n\
You should  have received  a copy of the GNU  General  Public License along with\n\
this program.  If not, see <http://www.gnu.org/licenses/>,  or write to the Free\n\
Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the  U.S. Government  (\"Government\"),  the following provisions apply to you.\n\
If the Software is  supplied by the Department of Defense (\"DoD\"), it is classi-\n\
fied as  \"Commercial Computer Software\"  under paragraph 252.227-7014 of the DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the license rights  granted\n\
herein (the license  rights customarily  provided to non-Government  users).  If\n\
the Software is supplied to any unit or agency of the Government other than DoD,\n\
it is classified as  \"Restricted Computer Software\" and the  Government's rights\n\
in the  Software are defined in  paragraph 52.227-19 of the Federal  Acquisition\n\
Regulations  (\"FAR\") (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of the  NASA Supplement  to the  FAR (or  any  successor\n\
regulations).\n\
\n\
", argv[0], ident);
}

void
version(int argc, char *argv[])
{
	fprintf(stdout, "\
\n\
%1$s %2$s:\n\
    Copyright (c) 2003-2007  OpenSS7 Corporation.  All Rights Reserved.\n\
    Copyright (c) 1997       David Grothe, Gcom, Inc <dave@gcom.com>\n\
\n\
    Distributed by OpenSS7 Corporation under GPL Version 3, included\n\
    here by reference.\n\
\n\
    See `%1$s --copying' for copying permissions.\n\
\n\
", argv[0], ident);
}

void
usage(int argc, char *argv[])
{
	fprintf(stderr, "\
Usage:\n\
    %1$s [options] [{-c,--count}] count\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
", argv[0]);
}

void
help(int argc, char *argv[])
{
	fprintf(stdout, "\
\n\
Usage:\n\
    %1$s [options] [{-c,--count}] count\n\
    %1$s {-h,--help}\n\
    %1$s {-V,--version}\n\
    %1$s {-C,--copying}\n\
Options:\n\
    [-c, --count] COUNT\n\
        sets the iteration count to COUNT [default: %2$ld]\n\
    -q, --quiet\n\
        suppresses normal output\n\
    -h, --help\n\
        print this usage information and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
\n\
", argv[0], iter_cnt);
}

int
main(int argc, char *argv[])
{
	/* 
	 *  Automake distcheck (reasonably) does not permit installed binaries
	 *  that do not accept --help and --version options.
	 */
	while (1) {
		int c;

#ifdef _GNU_SOURCE
		int option_index = 0;
	/* *INDENT-OFF* */
	static struct option long_options[] = {
	    { "count",	 required_argument, NULL, 'c' },
	    { "quiet",	 no_argument,	    NULL, 'q' },
	    { "help",	 no_argument,	    NULL, 'h' },
	    { "version", no_argument,	    NULL, 'V' },
	    { "copying", no_argument,	    NULL, 'C' },
	    { 0, }
	};
	/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "c:qhVC", long_options, &option_index);
#else				/* _GNU_SOURCE */
		c = getopt(argc, argv, "c:qhVC");
#endif				/* _GNU_SOURCE */

		if (c == -1)
			break;
		switch (c) {
		case 0:
			usage(argc, argv);
			exit(2);
		case 'q':
			output = 0;
			break;
		case 'h':
			help(argc, argv);
			exit(0);
		case 'V':
			version(argc, argv);
			exit(0);
		case 'C':
			copying(argc, argv);
			exit(0);
		case 'c':
			if (sscanf(optarg, "%li", &iter_cnt) != 1)
				goto bad_option;
			break;
		case '?':
		default:
		      bad_option:
			optind--;
		      bad_nonoption:
			if (optind < argc && output) {
				fprintf(stderr, "%s: illegal syntax -- ", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, "%s ", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
	}
	/* 
	 * mimic old behavior of taking iteration count as non-option argument 
	 */
	if (optind < argc) {
		if (sscanf(optarg, "%li", &iter_cnt) != 1)
			goto bad_nonoption;
		optind++;
	}
	/* 
	 * don't ignore additional (permuted) non-option arguments 
	 */
	if (optind < argc)
		goto bad_nonoption;

#if	!defined(LINUX) && !defined(QNX)
	register_drivers();
	make_nodes();
#endif

	version(argc, argv);
	copying(argc, argv);

	if (output) {
		printf("Timing test version %s\n\n", "2.5 09/02/04");
		printf("Using safe constructs and message tracing\n");
	}
	set_debug_mask(debug_mask | 0x30000L);
	timing_test();

	if (output)
		printf("\n\nNot using safe constructs or message tracing\n");
	set_debug_mask(debug_mask & ~0x30000L);
	timing_test();

	return 0;
}
